matmul.hip

```diff
@@ -2657,6 +2657,150 @@ class MmtKernel_256t_MSxNS_amdgcn_mfma_i32_16x16x32i8_shared_Kx2_pipeline_v3
+                                m_outer * K_outer * A_tile_size_in_vec16;
+    const int64x2_t *B_global = static_cast<const int64x2_t *>(B_data) +
+                                n_outer * K_outer * B_tile_size_in_vec16;
+
+    __shared__ int64x2_t A_shared[A_tile_size_in_vec16];
+    __shared__ int64x2_t B_shared[B_tile_size_in_vec16];
+
+    int64x2_t A_vgpr0[A_tile_size_in_vec16 / num_threads];
+    int64x2_t B_vgpr0[B_tile_size_in_vec16 / num_threads];
+
+    int64x2_t A_block_vgpr1[MS];
+    int64x2_t B_block_vgpr1[NS/2];
+
+    auto vpgr0_to_shared = [&]() {
+      for (int i = 0; i < A_tile_size_in_vec16 / num_threads; ++i) {
+        A_shared[i * num_threads + tid] = A_vgpr0[i];
+      }
+      for (int j = 0; j < B_tile_size_in_vec16 / num_threads; ++j) {
+        B_shared[j * num_threads + tid] = B_vgpr0[j];
+      }
+    };
+
+    auto global_to_shared = [&]() {
+      for (int i = 0; i < A_tile_size_in_vec16 / num_threads; ++i) {
+        __builtin_amdgcn_global_load_lds(&A_global[i * num_threads + tid], &A_shared[i * num_threads + tid], 4, 0, 0);
+        __builtin_amdgcn_global_load_lds(&A_global[i * num_threads + tid], &A_shared[i * num_threads + tid], 4, 4, 0);
+        __builtin_amdgcn_global_load_lds(&A_global[i * num_threads + tid], &A_shared[i * num_threads + tid], 4, 8, 0);
+        __builtin_amdgcn_global_load_lds(&A_global[i * num_threads + tid], &A_shared[i * num_threads + tid], 4, 12, 0);
+      }
+      for (int j = 0; j < B_tile_size_in_vec16 / num_threads; ++j) {
+        __builtin_amdgcn_global_load_lds(&B_global[j * num_threads + tid], &B_shared[j * num_threads + tid], 4, 0, 0);
+        __builtin_amdgcn_global_load_lds(&B_global[j * num_threads + tid], &B_shared[j * num_threads + tid], 4, 4, 0);
+        __builtin_amdgcn_global_load_lds(&B_global[j * num_threads + tid], &B_shared[j * num_threads + tid], 4, 8, 0);
+        __builtin_amdgcn_global_load_lds(&B_global[j * num_threads + tid], &B_shared[j * num_threads + tid], 4, 12, 0);
+      }
+      A_global += A_tile_size_in_vec16;
+      B_global += B_tile_size_in_vec16;
+    };
+
+    auto shared_to_vgpr = [&]() {
+      for (int i = 0; i < MS; ++i)
+        A_block_vgpr1[i] = A_shared[64 * i + (tid % 64)];
+      for (int j = 0; j < NS / 4; ++j)
+        B_block_vgpr1[j] = B_shared[256 * j + tid];
+    };
+
+    auto mfma = [&]() {
+      for (int i = 0; i < MS; ++i) {
+        for (int j = 0; j < NS / 4; ++j) {
+          for (int k = 0; k < 2; ++k) {
+            acc[i][j] = __builtin_amdgcn_mfma_i32_16x16x32_i8(
+                A_block_vgpr1[i][k], B_block_vgpr1[j][k], acc[i][j], 0, 0, 0);
+          }
+        }
+      }
+    };
+
+    auto sync = [] {
+      __builtin_amdgcn_s_barrier();
+    };
+
+    global_to_shared();
+    if (K_outer >= 2) {
+      global_to_shared();
+      sync();
+      for (int k_outer = 0; k_outer < K_outer - 2; ++k_outer) {
+        global_to_shared();
+        mfma();
+        sync();
+        shared_to_vgpr();
```

Comment:
I don't understand this code. Why do we run `shared_to_vgpr` **after** the first matmul?

Reply:
after our discussion, I merged shared memory and let mfma to read from shared local on the C++ level.

I think this loses granularity which can benefit the software pipelining, performance could hurt.

Anyway, I will get some numbers first.